File: ios/Reguerta/Reguerta/Presentation/Access/MyOrderCommitmentValidation.swift
import Foundation

let ecoBasketOptionPickup = "pickup"
let ecoBasketOptionNoPickup = "no_pickup"

struct MyOrderCheckoutValidationResult: Equatable {
    let missingCommitmentProductNames: [String]
    let hasEcoBasketPriceMismatch: Bool

    var isValid: Bool {
        missingCommitmentProductNames.isEmpty && !hasEcoBasketPriceMismatch
    }
}

func validateMyOrderCheckout(
    currentMember: Member?,
    members: [Member],
    products: [Product],
    seasonalCommitments: [SeasonalCommitment] = [],
    selectedQuantities: [String: Int],
    selectedEcoBasketOptions: [String: String],
    currentWeekParity: ProducerParity = currentISOWeekProducerParity()
) -> MyOrderCheckoutValidationResult {
    let requiredEcoBasketProducts = requiredEcoBasketCommitmentProducts(
        currentMember: currentMember,
        members: members,
        products: products,
        currentWeekParity: currentWeekParity
    )
    let requiredSeasonalProducts = requiredSeasonalCommitmentProducts(
        products: products,
        seasonalCommitments: seasonalCommitments
    )

    let hasRequiredEcoBasket = requiredEcoBasketProducts.isEmpty || requiredEcoBasketProducts.contains { product in
        let quantity = selectedQuantities[product.id, default: 0]
        guard quantity > 0 else { return false }
        let selectedOption = selectedEcoBasketOptions[product.id]
        return selectedOption == ecoBasketOptionPickup || selectedOption == ecoBasketOptionNoPickup
    }

    let missingEcoNames = hasRequiredEcoBasket ? [] : requiredEcoBasketProducts.map(\.name)
    let missingSeasonalNames = requiredSeasonalProducts.compactMap { requirement -> String? in
        selectedQuantities[requirement.product.id, default: 0] < requirement.requiredUnits
            ? requirement.product.name
            : nil
    }
    let missingNames = deduplicatePreservingOrder(missingEcoNames + missingSeasonalNames)

    let distinctEcoBasketPrices = Set(
        products
            .filter(\.isEcoBasket)
            .filter(\.isVisibleInOrdering)
            .map(\.normalizedEcoBasketPriceKey)
    )

    return MyOrderCheckoutValidationResult(
        missingCommitmentProductNames: missingNames,
        hasEcoBasketPriceMismatch: distinctEcoBasketPrices.count > 1
    )
}

private func requiredEcoBasketCommitmentProducts(
    currentMember: Member?,
    members: [Member],
    products: [Product],
    currentWeekParity: ProducerParity
) -> [Product] {
    guard let member = currentMember,
          member.isActive,
          member.roles.contains(.member)
    else {
        return []
    }

    let ecoBasketProducts = products
        .filter(\.isEcoBasket)
        .filter(\.isVisibleInOrdering)

    guard !ecoBasketProducts.isEmpty else {
        return []
    }

    switch member.ecoCommitmentMode {
    case .weekly:
        return ecoBasketProducts
    case .biweekly:
        guard let parity = member.ecoCommitmentParity else {
            return ecoBasketProducts
        }
        guard parity == currentWeekParity else {
            return []
        }

        let eligibleProducerIds = Set(
            members
                .filter { producer in
                    producer.roles.contains(.producer) &&
                        producer.isActive &&
                        producer.producerCatalogEnabled &&
                        producer.producerParity == parity
                }
                .map(\.id)
        )

        return ecoBasketProducts.filter { product in
            eligibleProducerIds.contains(product.vendorId)
        }
    }
}

private struct SeasonalProductRequirement {
    let product: Product
    let requiredUnits: Int
}

private func requiredSeasonalCommitmentProducts(
    products: [Product],
    seasonalCommitments: [SeasonalCommitment]
) -> [SeasonalProductRequirement] {
    guard !seasonalCommitments.isEmpty else {
        return []
    }

    let visibleProductsById = Dictionary(uniqueKeysWithValues: products
        .filter(\.isVisibleInOrdering)
        .map { ($0.id, $0) })

    var requiredUnitsByProductId: [String: Int] = [:]

    for commitment in seasonalCommitments where commitment.active {
        guard visibleProductsById[commitment.productId] != nil else {
            continue
        }
        let requiredUnits = max(1, Int(ceil(commitment.fixedQtyPerOfferedWeek)))
        let existingUnits = requiredUnitsByProductId[commitment.productId] ?? 0
        requiredUnitsByProductId[commitment.productId] = max(existingUnits, requiredUnits)
    }

    return requiredUnitsByProductId
        .compactMap { productId, requiredUnits in
            guard let product = visibleProductsById[productId] else {
                return nil
            }
            return SeasonalProductRequirement(product: product, requiredUnits: requiredUnits)
        }
        .sorted { lhs, rhs in
            if lhs.product.companyName.localizedCaseInsensitiveCompare(rhs.product.companyName) != .orderedSame {
                return lhs.product.companyName.localizedCaseInsensitiveCompare(rhs.product.companyName) == .orderedAscending
            }
            return lhs.product.name.localizedCaseInsensitiveCompare(rhs.product.name) == .orderedAscending
        }
}

private func deduplicatePreservingOrder(_ values: [String]) -> [String] {
    var seen = Set<String>()
    return values.filter { seen.insert($0).inserted }
}

private extension Product {
    var normalizedEcoBasketPriceKey: Int {
        Int((price * 10_000).rounded())
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/presentation/access/MyOrderCommitmentValidation.kt
package com.reguerta.user.presentation.access

import com.reguerta.user.domain.access.EcoCommitmentMode
import com.reguerta.user.domain.access.Member
import com.reguerta.user.domain.access.ProducerParity
import com.reguerta.user.domain.access.isMember
import com.reguerta.user.domain.access.isProducer
import com.reguerta.user.domain.commitments.SeasonalCommitment
import com.reguerta.user.domain.products.Product
import com.reguerta.user.domain.products.ProductPricingMode
import java.text.Normalizer
import java.util.Locale
import kotlin.math.abs
import kotlin.math.roundToInt

internal const val EcoBasketOptionPickup = "pickup"
internal const val EcoBasketOptionNoPickup = "no_pickup"
private val CommitmentDiacriticMarksRegex = "\\p{Mn}+".toRegex()
private val CommitmentTokenRegex = "[a-z0-9]+".toRegex()

internal data class MyOrderCheckoutValidationResult(
    val missingCommitmentProductNames: List<String>,
    val exceededCommitmentProductNames: List<String>,
    val incompatibleCommitmentProductNames: List<String>,
    val hasEcoBasketPriceMismatch: Boolean,
) {
    val isValid: Boolean
        get() = missingCommitmentProductNames.isEmpty() &&
            exceededCommitmentProductNames.isEmpty() &&
            incompatibleCommitmentProductNames.isEmpty() &&
            !hasEcoBasketPriceMismatch
}

internal fun validateMyOrderCheckout(
    currentMember: Member?,
    members: List<Member>,
    products: List<Product>,
    seasonalCommitments: List<SeasonalCommitment> = emptyList(),
    selectedQuantities: Map<String, Int>,
    selectedEcoBasketOptions: Map<String, String>,
    currentWeekParity: ProducerParity = currentIsoWeekProducerParity(),
): MyOrderCheckoutValidationResult {
    val requiredEcoBasketProducts = requiredEcoBasketCommitmentProducts(
        currentMember = currentMember,
        members = members,
        products = products,
        currentWeekParity = currentWeekParity,
    )
    val requiredSeasonalProducts = requiredSeasonalCommitmentProducts(
        products = products,
        seasonalCommitments = seasonalCommitments,
    )

    val hasRequiredEcoBasket = requiredEcoBasketProducts.isEmpty() || requiredEcoBasketProducts.any { product ->
        val selectedQuantity = selectedQuantities[product.id].orZero
        if (selectedQuantity <= 0) {
            return@any false
        }
        val selectedOption = selectedEcoBasketOptions[product.id]
        selectedOption == EcoBasketOptionPickup || selectedOption == EcoBasketOptionNoPickup
    }
    val missingEcoNames = if (hasRequiredEcoBasket) {
        emptyList()
    } else {
        requiredEcoBasketProducts.map(Product::name)
    }

    val missingSeasonalNames = mutableListOf<String>()
    val exceededSeasonalNames = mutableListOf<String>()
    val incompatibleSeasonalNames = mutableListOf<String>()
    requiredSeasonalProducts.forEach { requirement ->
        if (!requirement.isRepresentableBySelectionStep) {
            incompatibleSeasonalNames.add(requirement.product.name)
            return@forEach
        }

        val selectedUnits = selectedQuantities[requirement.product.id].orZero
        val selectedQuantity = selectedUnits.toDouble() * requirement.selectionStep
        if (selectedQuantity + SeasonalCommitmentTolerance < requirement.requiredQuantity) {
            missingSeasonalNames.add(requirement.product.name)
            return@forEach
        }
        if (selectedQuantity - SeasonalCommitmentTolerance > requirement.requiredQuantity) {
            exceededSeasonalNames.add(requirement.product.name)
        }
    }

    val missingNames = (missingEcoNames + missingSeasonalNames).distinct()
    val exceededNames = exceededSeasonalNames.distinct()
    val incompatibleNames = incompatibleSeasonalNames.distinct()

    val distinctEcoBasketPrices = products
        .asSequence()
        .filter(Product::isEcoBasket)
        .filter(Product::isVisibleInOrdering)
        .map { product -> product.price.normalizedEcoBasketPriceKey() }
        .toSet()
    val hasPriceMismatch = distinctEcoBasketPrices.size > 1

    return MyOrderCheckoutValidationResult(
        missingCommitmentProductNames = missingNames,
        exceededCommitmentProductNames = exceededNames,
        incompatibleCommitmentProductNames = incompatibleNames,
        hasEcoBasketPriceMismatch = hasPriceMismatch,
    )
}

internal fun seasonalCommitmentUnitLimitsByProductId(
    products: List<Product>,
    seasonalCommitments: List<SeasonalCommitment>,
): Map<String, Int> = requiredSeasonalCommitmentProducts(
    products = products,
    seasonalCommitments = seasonalCommitments,
).mapNotNull { requirement ->
    if (!requirement.isRepresentableBySelectionStep) {
        return@mapNotNull null
    }
    val expectedUnits = requirement.requiredQuantity / requirement.selectionStep
    val requiredUnits = expectedUnits.roundToInt()
    if (requiredUnits > 0) {
        requirement.product.id to requiredUnits
    } else {
        null
    }
}.toMap()

private fun requiredEcoBasketCommitmentProducts(
    currentMember: Member?,
    members: List<Member>,
    products: List<Product>,
    currentWeekParity: ProducerParity,
): List<Product> {
    val member = currentMember ?: return emptyList()
    if (!member.isActive || !member.isMember) {
        return emptyList()
    }

    val ecoBasketProducts = products
        .filter(Product::isEcoBasket)
        .filter(Product::isVisibleInOrdering)
    if (ecoBasketProducts.isEmpty()) {
        return emptyList()
    }

    return when (member.ecoCommitmentMode) {
        EcoCommitmentMode.WEEKLY -> ecoBasketProducts
        EcoCommitmentMode.BIWEEKLY -> {
            val parity = member.ecoCommitmentParity ?: return ecoBasketProducts
            if (parity != currentWeekParity) {
                return emptyList()
            }
            val eligibleProducerIds = members
                .asSequence()
                .filter { producer ->
                    producer.isProducer &&
                        producer.isActive &&
                        producer.producerCatalogEnabled &&
                        producer.producerParity == parity
                }
                .map(Member::id)
                .toSet()

            ecoBasketProducts.filter { product ->
                eligibleProducerIds.contains(product.vendorId)
            }
        }
    }
}

private data class SeasonalProductRequirement(
    val product: Product,
    val requiredQuantity: Double,
    val selectionStep: Double,
    val isRepresentableBySelectionStep: Boolean,
)

private fun requiredSeasonalCommitmentProducts(
    products: List<Product>,
    seasonalCommitments: List<SeasonalCommitment>,
): List<SeasonalProductRequirement> {
    if (seasonalCommitments.isEmpty()) {
        return emptyList()
    }

    val visibleProducts = products
        .asSequence()
        .filter(Product::isVisibleInOrdering)
        .sortedWith(compareBy<Product> { it.companyName.lowercase() }.thenBy { it.name.lowercase() })
        .toList()
    val visibleProductsById = visibleProducts.associateBy(Product::id)
    val visibleProductsByName = linkedMapOf<String, Product>().apply {
        visibleProducts.forEach { product ->
            putIfAbsent(product.name.commitmentNormalized(), product)
        }
    }
    val visibleProductsWithNormalizedName = visibleProducts.map { product ->
        product to product.name.commitmentNormalized()
    }

    val requiredQuantityByProductId = seasonalCommitments
        .asSequence()
        .filter(SeasonalCommitment::active)
        .mapNotNull { commitment ->
            val matchedProductId = visibleProductsById[commitment.productId]?.id
                ?: commitment.productNameHint
                    ?.commitmentNormalized()
                    ?.let { visibleProductsByName[it]?.id }
                ?: commitment.commitmentSearchTerms().asSequence()
                    .mapNotNull { term ->
                        visibleProductsWithNormalizedName.firstOrNull { (_, normalizedName) ->
                            normalizedName.contains(term) || term.contains(normalizedName)
                        }?.first?.id
                    }
                    .firstOrNull()
                ?: return@mapNotNull null
            matchedProductId to commitment.fixedQtyPerOfferedWeek
        }
        .groupBy(
            keySelector = { it.first },
            valueTransform = { it.second },
        )
        .mapValues { (_, quantities) ->
            quantities.maxOrNull() ?: 0.0
        }

    return requiredQuantityByProductId
        .mapNotNull { (productId, requiredQuantity) ->
            if (requiredQuantity <= 0.0) {
                return@mapNotNull null
            }
            val product = visibleProductsById[productId] ?: return@mapNotNull null
            val selectionStep = product.commitmentSelectionStep()
            val expectedUnits = requiredQuantity / selectionStep
            val isRepresentableBySelectionStep =
                abs(expectedUnits - expectedUnits.roundToInt().toDouble()) <= SeasonalCommitmentTolerance
            SeasonalProductRequirement(
                product = product,
                requiredQuantity = requiredQuantity,
                selectionStep = selectionStep,
                isRepresentableBySelectionStep = isRepresentableBySelectionStep,
            )
        }
        .sortedWith(compareBy<SeasonalProductRequirement> { it.product.companyName.lowercase() }.thenBy { it.product.name.lowercase() })
}

private fun Product.commitmentSelectionStep(): Double =
    if (pricingMode == ProductPricingMode.WEIGHT) {
        unitQty.takeIf { it > 0.0 } ?: 1.0
    } else {
        1.0
    }

private fun Double.normalizedEcoBasketPriceKey(): String =
    String.format(Locale.US, "%.4f", this)

private fun String.commitmentNormalized(): String =
    Normalizer.normalize(trim(), Normalizer.Form.NFD)
        .replace(CommitmentDiacriticMarksRegex, "")
        .lowercase(Locale.getDefault())

private fun SeasonalCommitment.commitmentSearchTerms(): List<String> =
    listOfNotNull(productNameHint, seasonKey, productId)
        .flatMap { value ->
            CommitmentTokenRegex.findAll(value.commitmentNormalized())
                .map(MatchResult::value)
                .filter { token -> token.length >= 4 && token.any(Char::isLetter) }
                .toList()
        }
        .plus(seasonKey.commitmentSeasonAliasTerms())
        .distinct()

private fun String.commitmentSeasonAliasTerms(): List<String> {
    val firstToken = CommitmentTokenRegex.findAll(commitmentNormalized())
        .map(MatchResult::value)
        .firstOrNull()
        ?: return emptyList()

    return when (firstToken) {
        "avo", "avocado", "avocados", "aguacate", "aguacates" -> listOf(
            "aguacate",
            "aguacates",
            "avocado",
            "avocados",
        )
        "man", "mango", "mangos" -> listOf("mango", "mangos")
        else -> emptyList()
    }
}

private const val SeasonalCommitmentTolerance = 0.0001

private val Int?.orZero: Int
    get() = this ?: 0
